meta_client_v2/src/meta_impl.rs

```diff
@@ -0,0 +1,513 @@
+        Ok(GrpcClient {
+            client,
+            heartbeat_channel: RwLock::new(NodeHeartbeatChannel {
+                heartbeat_sender: sender,
+                action_cmd_receiver: Some(receiver),
+            }),
+        })
+    }
+
+    async fn reconnect_heartbeat_channel(&self) {
+        info!("gRPC reconnect begin");
+
+        loop {
+            match self.connect_grpc_client() {
+                Ok(client) => {
+                    *self.grpc_client.write().await = Some(client);
+
+                    info!("gRPC reconnect succeeds");
+                    return;
+                }
+                Err(e) => {
+                    error!("gRPC reconnect failed, err:{}", e);
+
+                    time::sleep(self.error_wait_lease()).await;
+                }
+            }
+        }
+    }
+
+    fn build_rpc_client(&self) -> CeresmetaRpcServiceClient {
+        let cb = ChannelBuilder::new(Arc::new(Environment::new(self.meta_config.cq_count)));
+        CeresmetaRpcServiceClient::new(cb.connect(&self.meta_config.meta_addr))
+    }
+
+    // TODO: Store the value in field
+    fn error_wait_lease(&self) -> Duration {
+        Duration::from_secs(self.meta_config.lease.as_secs() / 2)
+    }
+
+    async fn fetch_action_cmd_loop(&self, mut stop_rx: Receiver<()>) {
+        info!("Begin fetching action cmd loop");
+        loop {
+            debug!("Begin fetching action cmd once");
+            let receiver = match &*self.grpc_client.read().await {
+                Some(client) => client
+                    .heartbeat_channel
+                    .write()
+                    .await
+                    .action_cmd_receiver
+                    .take(),
+                None => {
+                    warn!("gRPC client is not inited when starting fetch action cmd");
+                    None
+                }
+            };
+
+            if let Some(v) = receiver {
+                match self.fetch_action_cmd(v).await {
+                    Ok(()) => {
+                        info!(
+                            "Fetch cluster action cmd finished, cluster:{}",
+                            self.get_cluster_name()
+                        );
+                    }
+                    Err(err) => {
+                        error!(
+                            "Failed to get action cmd, cluster:{}, err:{}",
+                            self.get_cluster_name(),
+                            err
+                        );
+                        self.reconnect_heartbeat_channel().await;
+                    }
+                }
+            } else {
+                warn!("Skip action command fetch because no receiver found");
+            }
+
+            if time::timeout(self.error_wait_lease(), stop_rx.recv())
+                .await
+                .is_ok()
+            {
+                warn!("Receive exit message, exit the fetch_action_cmd loop");
+                break;
+            }
```

Comment:
Is this reachable only after an existing command receiver is disconnected? To be more elegant I think we need to select over `stop_rx` and `self.fetch_action_cmd`

Reply:
It's my fault and actually the `fetch_action_cmd` is the real loop to fetch heartbeats...

Reply:
I will address it.